lib/watchers/fs-watcher.js

```diff
@@ -0,0 +1,439 @@
+                            cb();
+                            return;
+                        }
+                        cb(e);
+                        return;
+                    }
+
+                    self.getConfigTimestamp(vm,
+                        function _getConfigTimestampCb(error, latestCfg) {
+                            if (error) {
+                                cb(error);
+                                return;
+                            }
+                            // Thanks Javascript... NaN > X, where X is a
+                            // number. So we '|| 0' here to work around it.
+                            results[vm] = Math.max(st.mtime.getTime() || 0,
+                                latestCfg || 0);
+                            cb();
+                        }
+                    );
+                });
+            }
+        }, function _afterStats(e) {
+            var doneLookup = (new Date()).getTime();
+
+            self.log.trace({
+                action: 'getCurrentZones',
+                elapsed: (doneLookup - startLookup),
+                err: e,
+                vmCount: (vms ? Object.keys(vms).length : 0)
+            }, 'completed lookup of latest timestamps');
+
+            callback(e, results);
+        });
+    });
+};
+
+FsWatcher.prototype.getConfigTimestamp = // eslint-disable-line
+function getConfigTimestamp(vm, callback) {
+    var self = this;
+    var timestamps = [];
+
+    vasync.forEachParallel({
+        inputs: [
+            '/zones/' + vm + '/config/metadata.json',
+            '/zones/' + vm + '/config/routes.json',
+            '/zones/' + vm + '/config/tags.json'
+        ],
+        func: function _statConfigFile(filename, cb) {
+            fs.stat(filename, function _onStat(e, st) {
+                if (!e) {
+                    timestamps.push(st.mtime.getTime());
+                } else if (e.code !== 'ENOENT') {
+                    // If stat has an error we'll just log it since there's not
+                    // much else for us to do about it.
+                    self.log.warn(e, 'fs.stat error');
+                }
+                cb();
+            });
+        }
+    }, function _afterConfigFileStats(err) {
+        var latest = timestamps.sort().pop();
+
+        callback(err, latest);
+    });
+};
+
+/*
+ * For config we only watch for change events because when a VM is created we'll
+ * always see the .xml file but we may or may not see config. Since xml is
+ * always there we'll just rely on that.
+ *
+ */
+FsWatcher.prototype.watchConfig = function watchConfig(vmUuid, callback) {
+    var self = this;
+    var path;
+
+    if (self.configWatchers.hasOwnProperty(vmUuid)) {
+        // XXX can we tell if it's actually watching?
+        callback();
+        return;
+    }
+
+    path = '/zones/' + vmUuid + '/config';
+    try {
+        self.configWatchers[vmUuid] = // eslint-disable-line
+        fs.watch(path, function _watchHandler(evt) {
```

Comment:
I find not indenting on continuations a bit confusing, can we do that and add that rule to the `.eslintrc` file?